File: src/main/java/com/fengsheng/card/PingHeng.java
package com.fengsheng.card;

import com.fengsheng.*;
import com.fengsheng.phase.MainPhaseIdle;
import com.fengsheng.phase.OnUseCard;
import com.fengsheng.protos.Common;
import com.fengsheng.protos.Fengsheng;
import org.apache.log4j.Logger;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.TimeUnit;

public class PingHeng extends AbstractCard {
    private static final Logger log = Logger.getLogger(PingHeng.class);

    public PingHeng(int id, Common.color[] colors, Common.direction direction, boolean lockable) {
        super(id, colors, direction, lockable);
    }

    public PingHeng(int id, AbstractCard card) {
        super(id, card);
    }

    @Override
    public Common.card_type getType() {
        return Common.card_type.Ping_Heng;
    }

    @Override
    public boolean canUse(Game g, Player r, Object... args) {
        if (!(g.getFsm() instanceof MainPhaseIdle fsm) || r != fsm.player()) {
            log.error("平衡的使用时机不对");
            return false;
        }
        Player target = (Player) args[0];
        if (r == target) {
            log.error("平衡不能对自己使用");
            return false;
        }
        if (!target.isAlive()) {
            log.error("目标已死亡");
            return false;
        }
        return true;
    }

    @Override
    public void execute(final Game g, final Player r, Object... args) {
        final Player target = (Player) args[0];
        log.info(r + "对" + target + "使用了" + this);
        r.deleteCard(this.id);
        Fsm resolveFunc = () -> {
            for (Player player : g.getPlayers()) {
                if (player instanceof HumanPlayer p) {
                    p.send(Fengsheng.use_ping_heng_toc.newBuilder()
                            .setPlayerId(p.getAlternativeLocation(r.location())).setTargetPlayerId(p.getAlternativeLocation(target.location()))
                            .setPingHengCard(this.toPbCard()).build());
                }
            }
            g.playerDiscardCard(r, r.getCards().values().toArray(new Card[0]));
            g.playerDiscardCard(target, target.getCards().values().toArray(new Card[0]));
            r.draw(3);
            target.draw(3);
            g.getDeck().discard(this);
            return new ResolveResult(new MainPhaseIdle(r), true);
        };
        g.resolve(new OnUseCard(r, r, target, this, r, resolveFunc));
    }

    @Override
    public String toString() {
        return Card.cardColorToString(colors) + "平衡";
    }

    public static boolean ai(MainPhaseIdle e, Card card) {
        Player player = e.player();
        if (player.getCards().size() > 3) return false;
        List<Player> players = new ArrayList<>();
        for (Player p : player.getGame().getPlayers())
            if (p != player && p.isAlive()) players.add(p);
        if (players.isEmpty()) return false;
        Player p = players.get(ThreadLocalRandom.current().nextInt(players.size()));
        GameExecutor.post(player.getGame(), () -> card.execute(player.getGame(), player, p), 2, TimeUnit.SECONDS);
        return true;
    }
}

File: src/main/kotlin/card/PingHeng.kt
package com.fengsheng.card

import com.fengsheng.*
import com.fengsheng.phase.MainPhaseIdle
import com.fengsheng.phase.OnFinishResolveCard
import com.fengsheng.phase.ResolveCard
import com.fengsheng.protos.Common.card_type.Ping_Heng
import com.fengsheng.protos.Common.color
import com.fengsheng.protos.Common.direction
import com.fengsheng.protos.usePingHengToc
import com.fengsheng.skill.ConvertCardSkill
import com.fengsheng.skill.cannotPlayCard
import org.apache.logging.log4j.kotlin.logger
import java.util.concurrent.TimeUnit
import kotlin.math.abs

class PingHeng : Card {
    constructor(id: Int, colors: List<color>, direction: direction, lockable: Boolean) :
        super(id, colors, direction, lockable)

    constructor(id: Int, card: Card) : super(id, card)

    /**
     * 仅用于“作为平衡使用”
     */
    internal constructor(originCard: Card) : super(originCard)

    override val type = Ping_Heng

    override fun canUse(g: Game, r: Player, vararg args: Any): Boolean {
        if (r.cannotPlayCard(type)) {
            logger.error("你被禁止使用平衡")
            r.sendErrorMessage("你被禁止使用平衡")
            return false
        }
        if (r !== (g.fsm as? MainPhaseIdle)?.whoseTurn) {
            logger.error("平衡的使用时机不对")
            r.sendErrorMessage("平衡的使用时机不对")
            return false
        }
        val target = args[0] as Player
        if (r === target) {
            logger.error("平衡不能对自己使用")
            r.sendErrorMessage("平衡不能对自己使用")
            return false
        }
        if (!target.alive) {
            logger.error("目标已死亡")
            r.sendErrorMessage("目标已死亡")
            return false
        }
        return true
    }

    override fun execute(g: Game, r: Player, vararg args: Any) {
        val target = args[0] as Player
        val fsm = g.fsm as MainPhaseIdle
        logger.info("${r}对${target}使用了$this")
        r.deleteCard(id)
        val resolveFunc = { _: Boolean ->
            g.players.send {
                usePingHengToc {
                    playerId = it.getAlternativeLocation(r.location)
                    targetPlayerId = it.getAlternativeLocation(target.location)
                    pingHengCard = toPbCard()
                }
            }
            if (r.cards.isNotEmpty()) r.game!!.addEvent(DiscardCardEvent(r, r))
            if (target.cards.isNotEmpty()) r.game!!.addEvent(DiscardCardEvent(r, target))
            g.playerDiscardCard(r, r.cards.toList())
            g.playerDiscardCard(target, target.cards.toList())
            r.draw(3)
            target.draw(3)
            OnFinishResolveCard(r, r, target, getOriginCard(), Ping_Heng, fsm)
        }
        g.resolve(ResolveCard(r, r, target, getOriginCard(), Ping_Heng, resolveFunc, fsm))
    }

    override fun toString(): String {
        return "${cardColorToString(colors)}平衡"
    }

    companion object {
        fun ai(e: MainPhaseIdle, card: Card, convertCardSkill: ConvertCardSkill?): Boolean {
            val player = e.whoseTurn
            !player.cannotPlayCard(Ping_Heng) || return false
            player.cards.size <= 3 || return false
            val identity = player.identity
            val p = player.game!!.players.filter {
                if (it === player || !it!!.alive) false
                else if (identity != color.Black && identity == it.identity) it.cards.size <= 3
                else it.cards.size >= 3
            }
            p.size >= 1 || return false
            val target = p.maxBy { abs(it!!.cards.size - 3.1) }
            GameExecutor.post(player.game!!, {
                convertCardSkill?.onConvert(player)
                card.asCard(Ping_Heng).execute(player.game!!, player, target!!)
            }, 3, TimeUnit.SECONDS)
            return true
        }
    }
}
